kore/src/Kore/Internal/SideCondition.hs

```diff
@@ -577,3 +620,234 @@ retractLocalFunction =
             InternalSet_ _ -> Just (Pair term1 term2)
             _          -> Nothing
     go _ _ = Nothing
+
+{- | Assumes a 'TermLike' to be defined. If not always defined,
+it will be stored in the `SideCondition` together with any subterms
+resulting from the implication that the original term is defined.
+
+For maps and sets: this will generate and store a minimal set
+of sub-collections from which the definedness of any sub-collection
+can be inferred.
+-}
+assumeDefined
+    :: forall variable
+    .  InternalVariable variable
+    => TermLike variable
+    -> SideCondition variable
+assumeDefined term =
+    assumeDefinedWorker term
+    & fromDefinedTerms
+  where
+    assumeDefinedWorker
+        :: TermLike variable
+        -> HashSet (TermLike variable)
+    assumeDefinedWorker term' =
+        case term' of
+            TermLike.And_ _ child1 child2 ->
+                let result1 = assumeDefinedWorker child1
+                    result2 = assumeDefinedWorker child2
+                 in asSet term' <> result1 <> result2
+            TermLike.App_ symbol children ->
+                checkFunctional symbol term'
+                <> foldMap assumeDefinedWorker children
+            TermLike.Ceil_ _ _ child ->
+                asSet term' <> assumeDefinedWorker child
+            TermLike.InternalList_ internalList ->
+                asSet term'
+                <> foldMap assumeDefinedWorker (internalListChild internalList)
+            TermLike.InternalMap_ internalMap ->
+                let definedElems =
+                        getDefinedElementsOfAc internalMap
+                    definedMaps =
+                        generateNormalizedAcs internalMap
+                        & HashSet.map TermLike.mkInternalMap
+                 in foldMap assumeDefinedWorker definedElems
+                    <> asSet term'
```

Comment:
Why use `asSet` here?

Reply:
`generateNormalizedAcs` will not add the collection itself to the resulting `HashSet`. I thought it was clearer to separate the two, especially since we should check if the input collection is always defined or not before adding it. Actually, do you think we should do the same for the subcollections as well?

Reply:
Sorry, I mean: Why does the collection itself need to be in the set at all? The set should have all the sub-terms required to decide that the whole collection is defined.

Reply:
Done. We don't add the collection itself anymore, and I added a special case for singleton collections in `isDefined`.